text/3323-restrictions.md

```diff
@@ -0,0 +1,466 @@
+   trait IDENTIFIER
+   GenericParams? ( : TypeParamBounds? )? WhereClause? {
+     InnerAttribute*
+     AssociatedItem*
+   }
+
++ImplRestriction :
++     impl
++   | impl ( crate )
++   | impl ( self )
++   | impl ( mod )
++   | impl ( super )
++   | impl ( in SimplePath )
++   | impl ( in mod )
+```
+
+Essentially, `mut` and `impl` have the same syntax as `pub`, just with a different keyword.
+
+Finally, a minor change is needed to the syntax for visibility to permit `pub(mod)`.
+
+```diff
+Visibility :
+     pub
+   | pub ( crate )
+   | pub ( self )
++  | pub ( mod )
+   | pub ( super )
+   | pub ( in SimplePath )
++  | pub ( in mod )
+```
+
+## Behavior
+
+The current behavior of `pub` is that `pub` makes something visible within the declared scope. If no
+scope is declared (such that it is just `pub`), then the item is visible everywhere. This behavior
+is preserved for `impl` and `mut`. When a restriction is used, the behavior is allowed only within
+the declared scope. When the keyword is used alone, the behavior is allowed everywhere. While in
+most cases the default visibility is private, `pub` is default in some cases, namely `enum`
+variants, `enum` fields, and `trait` items. `impl` and `mut` will have a consistent default: when
+omitted entirely, the scope is inherited from `pub`. This is both what is most convenient and is
+what is required for backwards compatibility with existing code.
+
+`pub(mod)` is equivalent to `pub(self)`; likewise for `impl(mod)` and `mut(mod)`.
+
+When an `ImplRestriction` is present, implementations of the associated trait are only permitted
+within the designated path. Any implementation of the trait outside this scope is a compile error.
+When a `MutRestriction` is present, mutable uses of the associated field are only permitted within
+the designated path. Any mutable use of the field outside the scope is a compile error. Further, a
+`struct`, `union`, or `enum` variant containing fields with an associated `MutRestriction` may not
+be constructed with `struct` expressions unless all fields are unrestricted in the present scope.
+This is the case even if the field is not directly declared, such as when functional record updates
+are used.
+
+## "Mutable use" in the compiler
+
+The concept of a "mutable use" [already exists][mutating use method] within the compiler. This
+catches all situations that are relevant here, including `ptr::addr_of_mut!`, `&mut`, and direct
+assignment to a field, while excluding interior mutability. As such, formal semantics of what
+constitutes a "mutable use" are not stated here.
+
+[mutating use method]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/visit/enum.PlaceContext.html#method.is_mutating_use
+
+## Interaction with `trait` aliases
+
+Trait aliases cannot be implemented. As such, there is no concern about compatibility between the
+`impl` restriction and `trait` aliases.
+
+# Drawbacks
+
+- Additional syntax for macros to handle
```

Comment:
You need to cover how precisely macros *should* handle it. I think this needs at least one new matcher, possibly a few. First, we definitely need a matcher for "what goes inside the parentheses after `mut` or `impl`", matching `self`, `crate`, `super`, `in path`, etc. And second, we *might* want a convenient matcher for "everything that can go before the name in a field" and "everything that can go before the name in a trait declaration", which types may use in place of "vis" for convenience.

As an alternative, I wonder if it *might* make sense to have these desugar to an attribute, and then macros that match and preserve attributes will Just Work. That would be a larger change, though.

Reply:
You also need a second discussion about macros, separate from matching this syntax: is mutability/implementability determined by span or by expansion site?

Visibility currently uses expansion site, not span. However, span-based visibility would be *incredibly* valuable for many users (to avoid having to export a `doc(hidden)` type for use by a macro that users can technically use without the macro). Taking that into account, I wonder if we might *want* span-based mutability and span-based implementability.